### PR TITLE
[Feature][Connector-Paimon] Support dynamic bucket splitting improves Paimon writing efficiency

### PR DESCRIPTION
**Pull request by hawk9821**
### Purpose of this pull request
Support dynamic bucket splitting improves Paimon writing efficiency

### Does this PR introduce _any_ user-facing change?
no

### How was this patch tested?
e2e: PaimonSinkDynamicBucketIT 
UT: PaimonBucketAssignerTest#bucketAssigner
e2e case: **PaimonSinkDynamicBucketIT#testPaimonBucketCountOnSparkAndFlink** ，because spark and Flink engine can not auto create paimon table on worker node in local file,  this e2e case work on local hdfs environment.  

### Check list

* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [ ] Update the [`release-note`](https://github.com/apache/seatunnel/blob/dev/release-note.md).

### CODE REVIEW

**Comment by Hisoka-X** [comment]
cc @dailai and @TaoZex

File: docs/zh/connector-v2/sink/Paimon.md [hunk 1]
```diff
@@ -30,21 +30,20 @@ libfb303-xxx.jar
...
-| table                       | 字符串 | 是    | -                            | 表名                                                                                                |
-| hdfs_site_path              | 字符串 | 否    | -                            | hdfs-site.xml文件路径                                                                                 |
-| schema_save_mode            | 枚举  | 否    | CREATE_SCHEMA_WHEN_NOT_EXIST | Schema保存模式                                                                                        |
-| data_save_mode              | 枚举  | 否    | APPEND_DATA                  | 数据保存模式                                                                                            |
-| paimon.table.primary-keys   | 字符串 | 否    | -                            | 主键字段列表，联合主键使用逗号分隔(注意：分区字段需要包含在主键字段中)                                                              |
-| paimon.table.partition-keys | 字符串 | 否    | -                            | 分区字段列表，多字段使用逗号分隔                                                                                  |
```
Comment (Hisoka-X) [inline docs/zh/connector-v2/sink/Paimon.md]: Why deleted this config?

File: seatunnel-translation/seatunnel-translation-flink/seatunnel-translation-flink-13/src/main/java/org/apache/seatunnel/translation/flink/sink/FlinkSinkWriterContext.java [hunk 2]
```diff
@@ -50,6 +50,11 @@ public int getIndexOfSubtask() {
         return writerContext.getSubtaskId();
     }
 
+    @Override
+    public int getNumberOfParallelSubtasks() {
```
Comment (dailai) [inline seatunnel-translation/seatunnel-translation-flink/seatunnel-translation-flink-13/src/main/java/org/apache/seatunnel/translation/flink/sink/FlinkSinkWriterContext.java:56]: Please do not return 0 directly, if this is not necessary to implement, please give a default implementation in the interface, do not implement it here.

File: seatunnel-translation/seatunnel-translation-flink/seatunnel-translation-flink-common/src/main/java/org/apache/seatunnel/translation/flink/sink/FlinkSinkWriterContext.java [hunk 3]
```diff
@@ -47,6 +47,11 @@ public int getIndexOfSubtask() {
         return writerContext.getSubtaskId();
     }
 
+    @Override
+    public int getNumberOfParallelSubtasks() {
```
Comment (dailai) [inline seatunnel-translation/seatunnel-translation-flink/seatunnel-translation-flink-common/src/main/java/org/apache/seatunnel/translation/flink/sink/FlinkSinkWriterContext.java:53]: ditto

**Comment by dailai** [comment]
Please retrigger the ci.

File: seatunnel-connectors-v2/connector-paimon/src/main/java/org/apache/seatunnel/connectors/seatunnel/paimon/sink/bucket/PaimonBucketAssigner.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+import org.apache.paimon.reader.RecordReader;
+import org.apache.paimon.reader.RecordReaderIterator;
+import org.apache.paimon.schema.TableSchema;
+import org.apache.paimon.table.FileStoreTable;
+import org.apache.paimon.table.Table;
+import org.apache.paimon.table.sink.RowPartitionKeyExtractor;
+
+import java.io.IOException;
+
+public class PaimonBucketAssigner {
```
Comment (dailai) [inline seatunnel-connectors-v2/connector-paimon/src/main/java/org/apache/seatunnel/connectors/seatunnel/paimon/sink/bucket/PaimonBucketAssigner.java:35]: Please add UT case for this class to vertify the logical.

**Comment by dailai** [comment]
Thinks @hawk9821 . Good job. I think your e2e case needs to be added to the case of multi-parallelism, the current case is all single parallelism. In this way, we can effectively verify whether the dynamic bucketing will change depending on the degree of parallelism of the job. Also, I think you should check the bucket count in every case instead of making a separate case. In addition, each of your cases should verify that the dynamic-bucket.target-row-num argument works as expected.

**Comment by dailai** [comment]
Why are there so many file changes. Maybe you're having some problems with your git operations. Please reopen a pr which must only has your commits. Then you can link this pr in new pr and close this pr.

**Comment by hawk9821** [comment]
> Why are there so many file changes. Maybe you're having some problems with your git operations. Please reopen a pr which must only has your commits. Then you can link this pr in new pr and close this pr.

Problems caused by rebase ,  resolved

**Comment by hawk9821** [comment]
> Thinks @hawk9821 . Good job. I think your e2e case needs to be added to the case of multi-parallelism, the current case is all single parallelism. In this way, we can effectively verify whether the dynamic bucketing will change depending on the degree of parallelism of the job. Also, I think you should check the bucket count in every case instead of making a separate case. In addition, each of your cases should verify that the dynamic-bucket.target-row-num argument works as expected.

get, add e2e case  **PaimonSinkDynamicBucketIT.testParallelismBucketCount**  **PaimonSinkDynamicBucketIT.testCDCParallelismBucketCount**

**Comment by dailai** [comment]
LGTM @Hisoka-X PTAL

**Review by dailai — APPROVED** [review]

File: docs/en/connector-v2/sink/Paimon.md [hunk 5]
```diff
@@ -43,7 +43,7 @@ libfb303-xxx.jar
...
 | paimon.table.primary-keys   | String | No       | -                            | Default comma-separated list of columns (primary key) that identify a row in tables.(Notice: The partition field needs to be included in the primary key fields) |
 | paimon.table.partition-keys | String | No       | -                            | Default comma-separated list of partition fields to use when creating tables.                                                                                    |
-| paimon.table.write-props    | Map    | No       | -                            | Properties passed through to paimon table initialization, [reference](https://paimon.apache.org/docs/master/maintenance/configurations/#coreoptions).            |
+| paimon.table.write-props    | Map    | No       | -                            | Properties passed through to paimon table initialization, [reference](https://paimon.apache.org/docs/0.8/maintenance/configurations/#coreoptions).               |
```
Comment (Hisoka-X) [inline docs/en/connector-v2/sink/Paimon.md]: Please revert this change.

File: docs/en/connector-v2/sink/Paimon.md [hunk 6]
```diff
@@ -241,6 +241,88 @@ sink {
 }
 ```
 
+### Single dynamic bucket table with write props of paimon，operates on the primary key table and bucket is -1
+
+#### core options：[reference](https://paimon.apache.org/docs/0.8/primary-key-table/data-distribution/#dynamic-bucket)
```
Comment (Hisoka-X) [inline docs/en/connector-v2/sink/Paimon.md]:
```suggestion
core options：[reference](https://paimon.apache.org/docs/master/primary-key-table/data-distribution/#dynamic-bucket)
```

File: docs/en/connector-v2/sink/Paimon.md [hunk 7]
```diff
@@ -241,6 +241,88 @@ sink {
...
+
+|              name              | type | required | default values |                  Description                   |
+|--------------------------------|------|----------|----------------|------------------------------------------------|
+| dynamic-bucket.target-row-num  | long | 是        | 2000000L       | controls the target row number for one bucket. |
+| dynamic-bucket.initial-buckets | int  | 否        |                | controls the number of initialized bucket.     |
```
Comment (Hisoka-X) [inline docs/en/connector-v2/sink/Paimon.md]: Change chinese to english.

File: docs/en/connector-v2/sink/Paimon.md [hunk 8]
```diff
@@ -241,6 +241,88 @@ sink {
 }
 ```
 
+### Single dynamic bucket table with write props of paimon，operates on the primary key table and bucket is -1
```
Comment (Hisoka-X) [inline docs/en/connector-v2/sink/Paimon.md]:
```suggestion
### Write to dynamic bucket table
Single dynamic bucket table with write props of paimon，operates on the primary key table and bucket is -1
```

File: docs/en/connector-v2/sink/Paimon.md [hunk 9]
```diff
@@ -241,6 +241,88 @@ sink {
...
+    password = "******"
+    table-names = ["seatunnel.role"]
+  }
+}
+
+sink {
+  Paimon {
+    catalog_name="seatunnel_test"
+    warehouse="file:///tmp/seatunnel/paimon/hadoop-sink/"
+    database="seatunnel"
```
Comment (Hisoka-X) [inline docs/en/connector-v2/sink/Paimon.md]: It's duplicated?

File: docs/en/connector-v2/sink/Paimon.md (hunk 8 above)
Comment (Hisoka-X) [reply docs/en/connector-v2/sink/Paimon.md]: Also please add comment for this feature only worked on zeta.

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-paimon-e2e/src/test/java/org/apache/seatunnel/e2e/connector/paimon/PaimonSinkDynamicBucketIT.java [hunk 10]
```diff
@@ -0,0 +1,358 @@
...
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import static org.awaitility.Awaitility.given;
+
+@DisabledOnContainer(
+        value = {},
+        type = {EngineType.SPARK, EngineType.FLINK},
+        disabledReason =
+                "Spark and Flink engine can not auto create paimon table on worker node in local file(e.g flink tm) by savemode feature which can lead error")
```
Comment (Hisoka-X) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-paimon-e2e/src/test/java/org/apache/seatunnel/e2e/connector/paimon/PaimonSinkDynamicBucketIT.java:76]: Could you share more detail info about the reason of this error?

File: seatunnel-api/src/main/java/org/apache/seatunnel/api/sink/SinkWriter.java [hunk 11]
```diff
@@ -92,6 +92,11 @@ interface Context extends Serializable {
         /** @return The index of this subtask. */
         int getIndexOfSubtask();
 
+        /** @return parallelism of this writer. */
+        default int getNumberOfParallelSubtasks() {
```
Comment (Hisoka-X) [inline seatunnel-api/src/main/java/org/apache/seatunnel/api/sink/SinkWriter.java:96]: I am looking for a way to get spark/flink to support both returning parallelism.

File: docs/en/connector-v2/sink/Paimon.md [hunk 12]
```diff
@@ -241,6 +241,49 @@ sink {
 }
 ```
 
+### Write to dynamic bucket table 
+
+Single dynamic bucket table with write props of paimon，operates on the primary key table and bucket is -1, only worked on zeta.
+
+#### core options：[reference](https://paimon.apache.org/docs/master/primary-key-table/data-distribution/#dynamic-bucket)
```
Comment (Hisoka-X) [inline docs/en/connector-v2/sink/Paimon.md]:
```suggestion
#### core options
Please [reference](https://paimon.apache.org/docs/master/primary-key-table/data-distribution/#dynamic-bucket)
```

**Review by Hisoka-X — APPROVED** [review]
LGTM if ci passes. Thanks @hawk9821

**Review by hailin0 — APPROVED** [review]
